testament/testament.nim

```diff
@@ -1220,155 +1497,908 @@ proc parseArgs(execState: var Execution, p: var OptParser): ParseCliResult =
+      of actionCompile, actionReject:
+        msg Undefined:
+          "testSpec count: " & $count & " expected: " & $expected
+      of actionRun:
+        discard # we'll have reported the compile action already
+
+      continue # don't actually run any tests
+
+    # handle skipping of disabled and known issues
+    case spec.err:
+      of reDisabled, reKnownIssue:
+        execState.legacyTestResults.addResult(
+          testRun, "", "", spec.err)
+
+        continue
+      else:
+        discard # keep processing
+
+    case action.kind:
+      of actionRun:
+        let
+          isJsTarget = target == targetJs
+          # specFile = execState.testSpecs[testId].file
+          exeFile = testRun.exeFile(testFile.file, execState.rootDir)
+          exeCmd =
+            if isJsTarget:
+              findNodeJs()
+            elif spec.useValgrind != disabled:
+              "valgrind"
+            else:
+              exeFile.dup(normalizeExe)
+          args =
+            if isJsTarget:
+              @["--unhandled-rejections=strict", exeFile]
+            elif spec.useValgrind != disabled:
+              let leakCheck =
+                if spec.useValgrind == leaking:
+                  "yes"
+                else:
+                  "no"
+
+              @["--error-exitcode=1", "--leak-check=" & leakCheck, exeFile]
+            else:
+              @[]
+          # REFACTOR remove string-based operatoins, use shell helper
+          # instead.
+          runCmd = map(@[exeCmd] & args, quoteShell).join(" ")
+
+        if testCmds.len == 0: # we've already processed its dependency
+          testCmds.add runCmd
+          cmdIdToActId.add actionId
+          cmdIdToActKind.add action.kind
+          cmdIdToInput.add testInput
+        else: # dependency is in the current batch, add to the next
+          nextTestCmds.add runCmd
+          nextCmdIdToActId.add actionId
+          nextCmdIdToActKind.add action.kind
+          nextCmdIdToInput.add testInput
+
+      of actionCompile, actionReject:
+        # add to this batch
+        testCmds.add compileCmd
+        cmdIdToActId.add actionId
+        cmdIdToActKind.add action.kind
+        cmdIdToInput.add testInput
+
+    let
+      lastActionId = testActions.len - 1
+      lastAction = actionId == lastActionId
+      currBatchFull = testCmds.len == batchSize
+      processOpts = {poStdErrToStdOut, poUsePath}
+
+    if currBatchFull or lastAction:
+      # REFACTOR this is an absolute abomination of an implementation, it
+      # should be refactored into something that uses 3-4 arguments, not
+      # fucking 15.
+      #
+      # IDEA a possible solution would be to introduce a "TestBatch" object
+      # which is populated with the required procedures. In this case part
+      # of the function above could also me moved elsewhere and turned into
+      # constructor-like procedure.
+      #
+      # NOTE each batch contains a fair bit of mutable state that gets
```

Comment:
Pre-batching isn't so easy as you have to shuffle things between batches all the time as you're executing, there is a reason why it works more like a sliding window right now.

The code creates the runnable batch and assumed next batch. Then after seeing the result of the current bacth it updates the assumed next batch, plus any related reporting, to do things like drop dependent actions that are no longer valid.

Batch might not be the right word to use universally and so it's giving the wrong impression, it's more like window of test runs and a window of current actions batch and potential next actions batch.